2d/include/pcl/2d/keypoint.h

```diff
@@ -45,10 +45,11 @@
 
 namespace pcl {
 
+template <typename ImageType>
 class Keypoint {
 private:
-  Edge edge_detection;
-  Convolution conv_2d;
+  Edge<ImageType, ImageType> edge_detection;
+  Convolution<ImageType> conv_2d;
```

Comment:
Can you perhaps address this change in another PR. I understand that this might be a fix, but it's tampering with the public API and those should changes should not be buried inside this PR.

Reply:
I will push the [2d] fixes directly (instead of waiting to bundle up some more modules) as soon as this is merged. I'll force push to remove these last 2 commits and add the one from your review, +1 if I can, I still don't get when a force push is disruptive.

Reply:
> I still don't get when a force push is disruptive.

When you modify the history and there are active conversations going on in that history.

Reply:
> I still don't get when a force push is disruptive.

If you change a file in a force push, the entire file will be marked as changed instead of the single incremental section you might have changed. If you're reviewing the implementation of a new algorithm in which reviewing one file takes a long time, this slows down considerably the review time. The best way to experience this is to actually be bitten by it personally :) If you want, you can start reviewing some of my PRs and I'll start force pushing by default so you get a feel for how annoying the review process becomes.

> I'll force push to remove these last 2 commits and add the one from your review, +1 if I can,

In this scenario, just create an incremental commit an incremental change just reverting the changes in this class. Remember that `git checkout <file>` allows you to do that easily. Besides rebase with master, there's usually no reason to ever force push before the reviews are complete. After the reviews, you can squash these together and it looks like the change was never made in the first place.

Reply:
I got the feeling it's super annoying after your first comment, it seemed evident and I stopped, but yeah I never reviewed PRs and it might be useful to start somewhere. What you're saying afterwards is to not use a `rebase` to `drop` the commits but use a `revert`, got it. Cheers!